sdk/src/main/scala/com/horizen/certificatesubmitter/strategies/WithKeyRotationStrategy.scala

```diff
@@ -0,0 +1,159 @@
+    val history = sidechainNodeView.history
+    val state: SidechainState = sidechainNodeView.state
+
+    val withdrawalRequests: Seq[WithdrawalRequestBox] = state.withdrawalRequests(status.referencedEpoch)
+
+    val btrFee: Long = getBtrFee(status.referencedEpoch)
+    val ftMinAmount: Long = getFtMinAmount(status.referencedEpoch)
+    val endEpochCumCommTreeHash = lastMainchainBlockCumulativeCommTreeHashForWithdrawalEpochNumber(history, status.referencedEpoch)
+    val sidechainId = params.sidechainId
+
+    val previousCertificateOption: Option[WithdrawalEpochCertificate] = state.certificate(status.referencedEpoch - 1)
+
+
+    val schnorrKeysSignaturesListBytes =  getSchnorrKeysSignaturesListBytes(state, status.referencedEpoch)
+
+    val signersPublicKeyWithSignatures = schnorrKeysSignaturesListBytes.schnorrSignersPublicKeysBytesList.zipWithIndex.map {
+      case (pubKey, pubKeyIndex) =>
+        (new SchnorrProposition(pubKey), status.knownSigs.find(info => info.pubKeyIndex == pubKeyIndex).map(_.signature))
+    }
+
+    CertificateDataWithKeyRotation(
+      status.referencedEpoch,
+      sidechainId,
+      withdrawalRequests,
+      endEpochCumCommTreeHash,
+      btrFee,
+      ftMinAmount,
+      signersPublicKeyWithSignatures,
+      schnorrKeysSignaturesListBytes,
+      previousCertificateOption,
+      CryptoLibProvider.thresholdSignatureCircuitWithKeyRotation.generateKeysRootHash(
+        scala.collection.JavaConverters.seqAsJavaList(params.signersPublicKeys.map(sp => sp.pubKeyBytes())),
+        scala.collection.JavaConverters.seqAsJavaList(params.mastersPublicKeys.map(sp => sp.pubKeyBytes())))
+    )
+  }
+
+  override def getMessageToSign(sidechainNodeView: View, referencedWithdrawalEpochNumber: Int): Try[Array[Byte]] = Try {
+    val history = sidechainNodeView.history
+    val state = sidechainNodeView.state
+
+    val withdrawalRequests: Seq[WithdrawalRequestBox] = state.withdrawalRequests(referencedWithdrawalEpochNumber)
+
+    val btrFee: Long = getBtrFee(referencedWithdrawalEpochNumber)
+    val ftMinAmount: Long = getFtMinAmount(referencedWithdrawalEpochNumber)
+
+    val endEpochCumCommTreeHash: Array[Byte] = lastMainchainBlockCumulativeCommTreeHashForWithdrawalEpochNumber(history, referencedWithdrawalEpochNumber)
+    val sidechainId = params.sidechainId
+
+    val actualKeys = getSchnorrKeysSignaturesList(getSchnorrKeysSignaturesListBytes(state, referencedWithdrawalEpochNumber))
+    val customFields = actualKeys.getUpdatedKeysRootHash.serializeFieldElement()
+
+    CryptoLibProvider.thresholdSignatureCircuitWithKeyRotation
+      .generateMessageToBeSigned(withdrawalRequests.asJava, sidechainId, referencedWithdrawalEpochNumber,
+        endEpochCumCommTreeHash, btrFee, ftMinAmount, util.Arrays.asList(customFields))
+  }
+
+  private def getSchnorrKeysSignaturesListBytes(state: SidechainState, referencedWithdrawalEpochNumber: Int): SchnorrKeysSignaturesListBytes = {
```

Comment:
This method needs to be refactored. Consider the fix in the state storage. Consider that SchnorrKeysSignaturesListBytes was refactored according to the suggestions in the comments there.

What we can do here:
1)` val prevCertifierKeys =  state.certifiersKeys(referencedWithdrawalEpochNumber - 1).getOrElse(<throw exception, because if we have no value, we do something completely wrong>)`
2) `val newCertifierKeys = state.certifiersKeys(referencedWithdrawalEpochNumber).getOrElse(<throw exception, because if we have no value, we do something completely wrong>)`
3) only for the different elements (prev key at index N != new key at index N) -> ask for keyRotationProof. If not found -> exception. Note: avoid yield, unzip3, it is not necessary here. Regular `for(i <- 1 until keysLength).map(...)` is enough, keep in mind that `state.certifiersKeys` contains `vector` type. If we have introduced it, let's use it then properly.

This will reduce number of read operation from the database and make the logic more strict. Currently we the "mute" the errors: `case None => IndexedSeq[Array[Byte]]()`

Reply:
Refactored, but unzip3 is very useful in this case to avoid 3 times loop